packages/ui/src/primitives/connection.tsx

```diff
@@ -3,20 +3,36 @@ import * as React from 'react';
 
 import { Spinner } from './spinner';
 
-export const Root = React.forwardRef<HTMLDivElement, React.HTMLAttributes<HTMLDivElement>>(function Root(
-  { children, className, ...props },
-  forwardedRef,
-) {
+function getColumnCount<T>(items: T[], maxCols: number = 6): number {
+  const numRows = Math.ceil(items.length / maxCols);
+  const itemsPerRow = Math.ceil(items.length / numRows);
+  const rows: T[][] = Array.from({ length: numRows }, () => []);
+  let currentArrayIndex = 0;
+  for (const item of items) {
+    rows[currentArrayIndex].push(item);
+    if (rows[currentArrayIndex].length === itemsPerRow) {
+      currentArrayIndex++;
+    }
+  }
+  return Math.max(...rows.map(row => row.length));
+}
+
+export const Root = React.forwardRef<
+  HTMLDivElement,
+  React.HTMLAttributes<HTMLDivElement> & {
+    columns?: number;
+  }
+>(function Root({ children, className, ...props }, forwardedRef) {
+  const cols = getColumnCount(React.Children.toArray(children));
   return (
     <div
       ref={forwardedRef}
       {...props}
       className={cn(
-        '[--cl-connection-gap:theme(spacing.2)]',
-        '[--cl-connection-cols:2]',
-        'flex gap-[--cl-connection-gap] flex-wrap items-center justify-center',
+        'flex items-center justify-center flex-wrap [--cl-connection-gap:theme(spacing.2)] gap-y-[--cl-connection-gap] -mx-[calc(var(--cl-connection-gap)/2)]',
```

Comment:
Perhaps a nit, but wondering if we should just use:

```suggestion
        'flex items-center justify-center flex-wrap [--cl-connection-gap:theme(spacing.2)] -m-[calc(var(--cl-connection-gap)/2)]',
```

and then use `p-` instead of `px-` within the child

Reply:
Sorry I meant at the usage level like this: https://play.tailwindcss.com/KiPqalm1Xa

The wrapping card body has `space-y-6` and the `<Connection.Root />` is a direct child with the negative margin applied. If we include the negative vertical margin, they'd be conflicting with each other.